### PR TITLE
Make native c10d_functional ops work with AOTInductor

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113057
* #113215
* __->__ #113735
* #113438

Summary:
- Revised `c10d_functional` ops to conform to https://github.com/pytorch/pytorch/tree/main/aten/src/ATen/native#func
- Modifed `get_cpp_op_schema()` to handle mutable args and aliasing returns

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/csrc/distributed/c10d/Functional.cpp [hunk 1]
```diff
@@ -66,10 +66,11 @@ c10d::ReduceOp to_reduce_op(const std::string& reduce_op) {
...
 
-at::Tensor all_reduce_(
-    at::Tensor input,
-    const std::string& reduce_op,
-    const std::string& group_name) {
+at::Tensor& all_reduce_(
+    at::Tensor& input,
+    std::string reduce_op,
+    // c10::string_view group_name,
+    std::string group_name) {
```
Comment (chenyang78) [inline torch/csrc/distributed/c10d/Functional.cpp:73]: Out of cuorisity, are these function signature changes necessary for making native c10d_functional ops work with AOTInductor? Thanks.

**Review by desertfire — APPROVED** [review]

**Review by desertfire — COMMENTED** [review]
I am not the best person to review the distributed part.

File: test/distributed/test_c10d_functional_native.py [hunk 2]
```diff
@@ -53,6 +59,11 @@ def device(self) -> torch.device:
         return torch.device(f"cuda:{self.rank}")
 
     def _init_process_group(self) -> None:
+        # Allow testing aoti after torch.compile
+        torch._inductor.config.triton.store_cubin = True
```
[the remark this answers is not in the document]
Comment (yifuwang) [reply test/distributed/test_c10d_functional_native.py:63]: Sorry I missed this comment earlier. Without this, I'd run into an error if I compile a function with inductor followed by an aot inductor compilation of the same function.

**Pull request closed without merging**